controllers/storagecluster/external_ocs.go

```diff
@@ -0,0 +1,248 @@
+		}
+		return reconcile.Result{}, err
+	}
+
+	if response.StorageConsumerUUID == "" || response.GrantedCapacity == "" {
+		err = fmt.Errorf("External-OCS:OnboardConsumer:response is empty")
+		r.Log.Error(err, "empty response")
+		return reconcile.Result{}, err
+	}
+
+	instance.Status.ExternalStorage.ConsumerID = response.StorageConsumerUUID
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return reconcile.Result{}, nil
+}
+
+// offboardConsumer makes an API call to the external storage provider cluster for offboarding
+func (r *StorageClusterReconciler) offboardConsumer(instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	_, err := r.externalClusterClient.OffboardConsumer(context.Background(), instance.Status.ExternalStorage.ConsumerID)
+	if err != nil {
+		if s, ok := status.FromError(err); ok {
+			r.logGrpcErrorAndReportEvent(instance, OffboardConsumer, err, s.Code())
+		}
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// updateConsumerCapacity makes an API call to the external storage provider cluster to update the capacity
+func (r *StorageClusterReconciler) updateConsumerCapacity(instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	response, err := r.externalClusterClient.UpdateCapacity(
+		context.Background(),
+		instance.Status.ExternalStorage.ConsumerID,
+		instance.Spec.ExternalStorage.RequestedCapacity.String())
+	if err != nil {
+		if s, ok := status.FromError(err); ok {
+			r.logGrpcErrorAndReportEvent(instance, UpdateCapacity, err, s.Code())
+		}
+		return reconcile.Result{}, err
+	}
+
+	if response.GrantedCapacity == "" {
+		err = fmt.Errorf("External-OCS:UpdateCapacity:response is empty")
+		r.Log.Error(err, "empty response")
+		return reconcile.Result{}, err
+	}
+
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return reconcile.Result{}, nil
+}
+
+// getExternalConfigFromProvider makes an API call to the external storage provider cluster for json blob
+func (r *StorageClusterReconciler) getExternalConfigFromProvider(instance *ocsv1.StorageCluster) ([]ExternalResource, reconcile.Result, error) {
+
+	response, err := r.externalClusterClient.GetStorageConfig(context.Background(), instance.Status.ExternalStorage.ConsumerID)
+	if err != nil {
+		if s, ok := status.FromError(err); ok {
+			r.logGrpcErrorAndReportEvent(instance, GetStorageConfig, err, s.Code())
+
+			// storage consumer is not ready yet, requeue after some time
+			if s.Code() == codes.Unavailable {
+				return nil, reconcile.Result{RequeueAfter: time.Second * 5}, nil
+			}
+		}
+
+		return nil, reconcile.Result{}, err
+	}
+
+	var externalResources []ExternalResource
+
+	for _, eResource := range response.ExternalResource {
+
+		data := map[string]string{}
+		err = json.Unmarshal(eResource.Data, &data)
+		if err != nil {
+			return nil, reconcile.Result{}, err
```

Comment:
Would suggest having `log.Error` to capture which resource (kind + name) had the problem with unmarshaling

Reply:
done